### ISSUE TITLE
A reset function that does not deallocate

### ISSUE THREAD

**Issue by TethysSvensson** [fitzgen/bumpalo#26]
I was wondering if it might be possible to write a variant of the reset function that does not deallocate memory.

Instead it would go through all the chunks and set `chunk.ptr` to `chunk.data`. Afterwards it would set `bump.current_chunk_footer` to `bump.all_chunk_footers`.

For this to make sense, we would have to modify either `alloc_layout_fast` or `alloc_layout_slow`, to check for additional chunks before allocating a new one.

We could for instance insert code similar to this in the beginning of `alloc_layout_slow`:

```rust

while let Some(next) = self.current_chunk_footer.get().as_ref().next.get() {
    self.current_chunk_footer.set(next);
    if let Some(ptr) = self.try_alloc_layout_fast(layout) {
        return ptr;
    }
}
```

Alternatively we could do a similar loop directly inside `try_alloc_layout_fast`.

What do you think? Would you accept a PR implementing this?

**For context:** I am looking into the possibility of implementing the [flatbuffers crate](https://github.com/google/flatbuffers/) on top of bumpalo. One of the core features of the library is the fact that once it has warmed up, it will not need to allocate additional memory to serialize further messages -- and this feature is not currently available in bumpalo.

**Comment by fitzgen** [comment]
I would prefer not to complicate the allocation algorithm.

We can still reach a warmed-up, steady state without modifying the allocation algorithm if during `reset` the `Bump` keeps the largest chunk it has allocated (aka the current chunk), rather than the smallest chunk (the current behavior). This means that there may be one extra `reset` iteration before the steady state is reached (compared to keeping all chunks) but should result in better cache locality and a simpler implementation.

Between that and the `with_capacity` constructor, would your needs be met?

**Comment by TethysSvensson** [comment]
Yes, I believe that would be fine.

However I am still looking into whether it is even feasible to use bumpalo as a backend flatbuffers. For various technical reasons, most current implementations allocate from the end of a buffer instead of the beginning. I am still thinking about whether you can avoid this without incurring a large runtime cost and/or cost in terms of API usability.

**Comment by TethysSvensson** [comment]
This was solved by #33.
